### PR TITLE
REST API: add functionality for saving Business Address in JPO

### PR DESCRIPTION
**Pull request by AnnaMag**
Depends on https://github.com/Automattic/jetpack/pull/8480.

Save Business Address in a sidebar widget during the JPO flow.

#### Changes proposed in this Pull Request:

* This patch extends the onboarding endpoint to inject and subsequently update Business Address info as a sidebar widget. It is not included in the current JPO version.

#### Testing instructions:
* Checkout current master in your local Calypso
* Checkout this branch on your JP sandbox
* Go to `https://YourJetpackSandbox.com/wp-admin/admin.php?page=jetpack&action=onboard&calypso_env=development`, where `YourJetpackSandbox.com` is the domain of your JP sandbox
* Skip through the flow till the `Add a business address` step
* Add input data and move to the `Next step`
* Go to `https://YourJetpackSandbox.com/wp-admin/widgets.php` and/or your site and verify you see the newly created BA widget in the list of widgets and a post sidebar
* Go back to `Add a business address` and modify the BA input
* Verify that changes apply on your JP site

**Note:** enabling widget module when not done so is a separate PR

### CODE REVIEW

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 1]
```diff
@@ -1041,11 +1041,85 @@ private function _process_onboarding( $data ) {
...
+		$active_sidebars = get_option( 'sidebars_widgets', array() );
+		$excluded_keys = array(
+			'wp_inactive_widgets',
+			'array_version',
+		);
+		foreach ( $excluded_keys as $key ) {
+			if ( isset( $active_sidebars[ $key ] ) ) {
+				unset( $active_sidebars[ $key ] );
+			}
+		}
```
Comment (eliorivero) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Since there are only two keys, the lines from sidebar option fetch up to the loop could be written as
```php
$active_sidebars = get_option( 'sidebars_widgets', array() );
unset( $active_sidebars[ 'wp_inactive_widgets' ], $active_sidebars[ 'array_version' ] );
```
The `unset` won't fail if the array item is not set.

It could also be written with an uglier one liner that ends up being longer :D
```php
$active_sidebars = array_diff_key(
    get_option( 'sidebars_widgets', array() ),
    array_flip( array( 'wp_inactive_widgets', 'array_version' ) )
);
```
or

Comment (AnnaMag) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Thanks for the review :) I settled for the first option to aid readability 👍

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 2]
```diff
@@ -1041,11 +1043,76 @@ private function _process_onboarding( $data ) {
...
 
+		public function get_first_sidebar() {
+			$active_sidebars = Jetpack_Widgets::get_active_sidebars();
+
+			unset( $active_sidebars[ 'wp_inactive_widgets' ], $active_sidebars[ 'array_version' ] );
+
+			if ( empty( $active_sidebars ) ) {
+				return false;
+			}
+			return array_shift( array_keys( $active_sidebars ) );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: This will throw a notice, because `array_shift` also updates the original referenced array (and there is no reference here):
```
PHP Notice:  Only variables should be passed by reference in .../jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php on line 1093
```

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 3]
```diff
@@ -1041,11 +1043,76 @@ private function _process_onboarding( $data ) {
...
 		}
 
+		// Add/update business address widget
+		if ( isset( $data['businessAddress'] ) ) {
+			//grab the first sidebar and proceed if it is present
+			$first_sidebar = self::get_first_sidebar();
+
+			if( $first_sidebar ) {
+				$title = wp_unslash( $data[ 'businessAddress' ][ 'name' ] );
+				$address = wp_unslash(
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Order seems a little weird to me - `city, state, street, zip`. In the original JPO plugin it was `street, city, state, zip` - https://github.com/Automattic/jetpack-onboarding/blob/master/class.jetpack-onboarding-end-points.php#L708.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 4]
```diff
@@ -1041,11 +1043,76 @@ private function _process_onboarding( $data ) {
 			}
 		}
 
+		// Add/update business address widget
+		if ( isset( $data['businessAddress'] ) ) {
+			//grab the first sidebar and proceed if it is present
+			$first_sidebar = self::get_first_sidebar();
+
+			if( $first_sidebar ) {
+				$title = wp_unslash( $data[ 'businessAddress' ][ 'name' ] );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Do we want to store `$data[ 'businessAddress' ]` in a variable to make the code a little shorter by accessing all those fields with less code?

Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Currently, we don't validate the presence of `name`, `street` or any of the others before using it; so we might end up using undeclared variables.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 3 above)
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Do we need those `wp_unslash`? Currently they might remove any backslashes that we have in the address.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 5]
```diff
@@ -1041,11 +1043,76 @@ private function _process_onboarding( $data ) {
...
+				$widget_options = array(
+					'title'   => $title,
+					'address' => $address,
+					'phone'   => '',
+					'hours'   => '',
+					'showmap' => false,
+					'email' => '',
+					'apikey' => '',
+				);
+				$position = '0';
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: The JPO plugin was appending the widget to the bottom of the sidebar. I think that makes more sense than prepending it before all widgets. We don't really want to tamper with the existing widgets.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 6]
```diff
@@ -1041,11 +1043,76 @@ private function _process_onboarding( $data ) {
...
+				$id_base = 'widget_contact_info';
+
+				$has_ba_widget = self::has_business_info_widget( $first_sidebar );
+				if ( $has_ba_widget ) {
+					self::update_widget( $id_base, $first_sidebar, $position, $widget_options);
+				}	else {
+					Jetpack_Widgets::activate_widget( $id_base, $first_sidebar, $position, $widget_options);
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Missing a space before the closing paren.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 7]
```diff
@@ -1041,11 +1043,76 @@ private function _process_onboarding( $data ) {
...
+				}
+			}
+		}
+
 		return empty( $error )
 			? ''
 			: join( ', ', $error );
 	}
 
+		public function get_first_sidebar() {
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: That method should be defined in `Jetpack_Widgets` instead.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 8]
```diff
@@ -1041,11 +1043,76 @@ private function _process_onboarding( $data ) {
...
+			$active_sidebars = Jetpack_Widgets::get_active_sidebars();
+
+			unset( $active_sidebars[ 'wp_inactive_widgets' ], $active_sidebars[ 'array_version' ] );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: We only want to unset those if they're already set. `Jetpack_Widgets::get_active_sidebars()` shouldn't contain a `wp_inactive_widgets` (I'm looking at it's code).

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 9]
```diff
@@ -1041,11 +1043,76 @@ private function _process_onboarding( $data ) {
...
+
+			if ( empty( $active_sidebars ) ) {
+				return false;
+			}
+			return array_shift( array_keys( $active_sidebars ) );
+		}
+
+		public function has_business_info_widget( $sidebar ) {
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Can we make this a method in `Jetpack_Widgets` that accepts a widget base ID (`widget_contact_info` in our case)? Could be useful, and would be the right place to keep this code IMHO.
 
Also, can we use something from `Jetpack_Widgets` to make this one simpler? `get_widget_by_id_base()` looks like a good candidate.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 10]
```diff
@@ -1041,11 +1043,76 @@ private function _process_onboarding( $data ) {
...
+			foreach ( $sidebars_widgets[ $sidebar ] as $widget ) {
+				if ( strpos( $widget, 'widget_contact_info' ) !== false ) {
+					return true;
+				}
+			}
+			return false;
+		}
+
+		public function update_widget( $id_base, $sidebar, $position, $widget_options ) {
+
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Unnecessary whitespace in the beginning of a block.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 11]
```diff
@@ -1041,11 +1043,76 @@ private function _process_onboarding( $data ) {
...
+			foreach ( $sidebars_widgets[ $sidebar ] as $widget ) {
+				if ( strpos( $widget, 'widget_contact_info' ) !== false ) {
+					return true;
+				}
+			}
+			return false;
+		}
+
+		public function update_widget( $id_base, $sidebar, $position, $widget_options ) {
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Should this method be in `_inc/lib/widgets.php` instead? Guess we should figure another name for it? `update_widget_by_id_base` maybe?

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 12]
```diff
@@ -1041,11 +1043,76 @@ private function _process_onboarding( $data ) {
...
+
+			$widget = Jetpack_Widgets::get_widget_by_id_base( $id_base );
+			$widget_id = $widget['id'];
+			Jetpack_Widgets::update_widget($widget_id,  $sidebar, $position, $widget_options );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Two consecutive spaces after `$widget_id,`.

File: _inc/lib/widgets.php [hunk 13]
```diff
@@ -120,6 +120,24 @@ public static function get_widget_by_id( $widget_id ) {
 		return $found;
 	}
 
+	/**
```
Comment (tyxla) [inline _inc/lib/widgets.php]: Left some feedback in the other PR about this one.

**Comment by AnnaMag** [comment]
Thanks for the reviews @tyxla 🙏 

I see that a lot is concerned around what to keep in the endpoint and what to move to module's methods. I kept a lot in the former, not being sure how much of that extends beyond JPO itself. I'll have another 👀 after we merge the previous related PRs.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 14]
```diff
@@ -1041,11 +1043,76 @@ private function _process_onboarding( $data ) {
...
+					'phone'   => '',
+					'hours'   => '',
+					'showmap' => false,
+					'email' => '',
+					'apikey' => '',
+				);
+				$position = '0';
+				$id_base = 'widget_contact_info';
+
+				$has_ba_widget = self::has_business_info_widget( $first_sidebar );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: @AnnaMag one thing to consider here: for the GET endpoint, we'll need to be able to:

1. Know whether business address has been set or not
2. Retrieve all those business address fields, in order to be able to return them in the endpoint.

Right now, with your PR, for 1, we can use `self::has_business_info_widget( $first_sidebar )`, but for 2. we'll need something else, for example storing `$data[ 'businessAddress' ]` in an option. If you decide to do that, make sure to also add the option to the list of known options, in order it to be cleaned up after uninstalling JP (as done here: https://github.com/Automattic/jetpack/pull/8495/files#diff-cb166c81692a76d1a92d53cb5e80d67cR537).

Comment (AnnaMag) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: As per Slack chat, let's handle it in the follow-up. As always 💯 point.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 8 above)
Comment (AnnaMag) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Good point about `wp_inactive_widgets`.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 1 above)
Comment (AnnaMag) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Using `Jetpack_Widgets` does not allow 'wp_inactive_widgets', so there is in fact a single key :)

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 9 above)
Comment (AnnaMag) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: @tyxla, I see the point, and at the same time thinking that it is pretty much a JPO-specific method. Will it be reusable by being defined in a Jetpack module? 🤔

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 11 above)
Comment (AnnaMag) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: I put it in a separate method when at the beginning of getting it up and running the code was much longer. Looking at it now, I reckon it is not needed in fact. Slightly re-written it becomes a two-liner and can be used in-place. Do you agree?

Comment (tyxla) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Not sure what exactly you mean - could you please elaborate?

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 9 above)
Comment (tyxla) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: No usages right now AFAIK, but it doesn't cost anything (other than just adding a parameter for the widget ID base) to do it, and in case we decide to use it in the future, it will be easier and will not require refactor PRs like this one #8464.

File: _inc/lib/widgets.php [hunk 15]
```diff
@@ -373,16 +387,23 @@ public static function set_widget_settings( $widget_id, $settings ) {
...
 		$instance_key = self::get_widget_instance_key( $widget_id );
-		$old_settings = $widget_settings[ $instance_key ];
 
-		if ( ! $settings = self::sanitize_widget_settings( $widget_id, $settings, $old_settings ) ) {
-			return new WP_Error( 'invalid_data', 'Update failed.', 500 );
-		}
-		if ( is_array( $old_settings ) ) {
-			// array_filter prevents empty arguments from replacing existing ones
-			$settings = wp_parse_args( array_filter( $settings ), $old_settings );
+		if ( ! $widget = self::get_registered_widget_object( self::get_widget_id_base( $widget_id ) ) {
```
Comment (oskosk) [inline _inc/lib/widgets.php]: There's a missing parens here

Comment (AnnaMag) [reply _inc/lib/widgets.php]: Was fixed in #8426 and now also ported here. Thanks for reviewing!

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 11 above)
Comment (AnnaMag) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: I removed that function and included: https://github.com/Automattic/jetpack/pull/8426/files#diff-2ccec9069213650359171727de9a2471R1073.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 16]
```diff
@@ -1041,10 +1043,59 @@ private function _process_onboarding( $data ) {
 			}
 		}
 
+		// Add/update business address widget
+		if ( isset( $data['businessAddress'] ) ) {
+			//grab the first sidebar and proceed if it is present
+			$first_sidebar = Jetpack_Widgets::get_first_sidebar();
```
Comment (ockham) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: This gives me 

> PHP Deprecated:  Non-static method Jetpack_Widgets::get_first_sidebar() should not be called statically in jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php on line 1049

Since `get_first_sidebar()` doesn't seem to refer to the instance ( `$this`), I think we can actually make it `static`.

**Comment by ockham** [comment]
I can't seem to get this to work :slightly_frowning_face: 

How I'm testing: Using the Jetpack Beta Plugin, and selecting 'Bleeding Edge' -- which at this point was 'last updated 4 hours ago', so it doesn't include #8501. Thus, I'm manually applying #8501 and this PR's diff.

The error I'm getting is the 'Unexpected Error' notice, and checking `/wp-admin/widgets.php`, I see that no widget was added. My browser's 'Network' tab gives me

```{"code":500,"headers":[{"name":"Content-Type","value":"application\/json"}],"body":{"error":"no_response_body","message":"Server could not read response."}}```

The server debug log contains (a [PHP Depracted](https://github.com/Automattic/jetpack/pull/8426/files#r161009496) and [PHP Notice](https://github.com/Automattic/jetpack/pull/8480/files#r160987585) aside), the following PHP Warning:

```PHP Warning:  Cannot modify header information - headers already sent by (output started at wp-content/plugins/jetpack-dev/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php:1069) in wp-includes/rest-api/class-wp-rest-server.php on line 1256```

**Comment by ockham** [comment]
Tried this now with simply checking out the `add/jpo-save-business-address` branch on my sandbox (per @AnnaMag's suggestion -- I was previously assuming I'd have to build JP for it to work), and making sure I'm using the corresponding `update/jpo-save-business-address` Calypso branch, but unfortunately, I'm still getting the same error :slightly_frowning_face:

**Comment by AnnaMag** [comment]
Closing in favor of https://github.com/Automattic/jetpack/pull/8523.
